README.rst
===============================
Dianonymous
===============================

A Python package for anonymizing DICOM files.  ***DISCLAIMER*** the
anonymization provided this package is not guaranteed to be complete! In
particular no effort is made to remove *burned-in* information from image
files.  Use at your own risk and you alone are responsible for ensuring your
patient data is properly anonymized.

This package "works for me" but MAY NOT FULLY ANONYMIZE YOUR DICOM FILES.

* Free software: BSD license

Features
--------

* TODO

Install
-------

This will install dianonymous in your Python packages directory
as well as a script called dianon in your Python scripts directory.  ::

    git clone git@github.com:randlet/dianonymous.git
    cd dianonymous
    python setup.py install

Run Tests
---------

Running the dianonymous tests ::

    git clone git@github.com:randlet/dianonymous.git
    cd dianonymous
    python setup.py nosetests

Usage
-----

From The Command Line
=====================

Anonymize a DICOM file and write to ./anonymize/ ::

    dianon path/to/file.dcm

Recursively anonymize all DICOM files in path/to/folder and write them to ./anonymize/ ::

    dianon -r path/to/folder

Recursively anonymize all DICOM files in path/to/folder and write them to different/output/path/ ::

    dianon -r -o different/output/path path/to/folder

Anonymize a DICOM file and write to ./anonymize/::

    dianon -r -o different/output/path path/to/folder


From Python
===========

From a Python script or interpreter ::

    import dianonymous
    paths_to_files_and_dirs = ["dir1/","dir2/","file1.dcm", "dir3/file2.dcm",...]
    output_dir = "./anonymized"

    dianonymous.anonymize(
        paths_to_files_and_dirs,
        output_dir,
        anon_id = None,    # optional string to use for output patient id
        anon_name = None,  # optional string to use for output patient name
        recurse = False,   # recurse any input subdirectories looking for dicom files
        private = True,    # delete any private tags
        curves = True,     # delete all curves
        log    = None,     # optional object with a .write method
    )



Acknowledgements
----------------

Ideas for the anonymizer were taken from both the pydicom and dicompyler packages


HISTORY.rst
.. :changelog:

History
-------

0.1.0 (2013-09-04)
++++++++++++++++++

* First release on PyPI.

tests/test_dianonymous.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_dianonymous
----------------------------------

Tests for `dianonymous` module.
"""

import dicom
import os
import shutil
import unittest

from dianonymous import dianonymous, anonymizer


class TestDianonymous(unittest.TestCase):

    def setUp(self):
        self.test_dir = os.path.join(os.path.dirname(__file__),"testfiles")
        self.anonymous_dir = os.path.join(os.path.dirname(__file__),"anonymized")
        if os.path.exists(self.anonymous_dir):
            shutil.rmtree(self.anonymous_dir)
        os.mkdir(self.anonymous_dir)

    def test_output_dir_exists(self):
        output = dianonymous.get_output_dir(self.anonymous_dir)
        self.assertEqual(output, self.anonymous_dir)

    def test_output_dir_doesnt_exist(self):
        out_dir = os.path.join(self.anonymous_dir,"doesntexist")
        output = dianonymous.get_output_dir(out_dir)
        self.assertEqual(output, out_dir)

    def test_input_paths_no_recurse(self):
        input_paths = os.listdir(self.test_dir)
        expected = set([i for i in input_paths if not os.path.isdir(i)])
        self.assertSetEqual(expected, set(dianonymous.get_input_paths(input_paths, False)))

    def test_input_paths_recurse(self):
        input_paths = []
        for root, _, files in  os.walk(self.test_dir):
            root = os.path.relpath(root)
            input_paths.extend(os.path.join(root,f) for f in files)

        expected = set(input_paths)
        self.assertSetEqual(expected, set(list(dianonymous.get_input_paths(self.test_dir, True))))

    def test_anonymize_no_id(self):
        dianonymous.anonymize(self.test_dir, recurse=True)

    def test_anonymize_with_id(self):
        dcm = dicom.read_file(os.path.join(self.test_dir,"rtplan.dcm"), force=True)
        anonymizer.anonymize(dcm, patient_id="testid")
        self.assertEqual(dcm.PatientID, "testid")

    def test_anonymize_with_name(self):
        dcm = dicom.read_file(os.path.join(self.test_dir,"rtplan.dcm"), force=True)
        anonymizer.anonymize(dcm, patient_name="testname")
        self.assertEqual(dcm.PatientsName, "testname")

    def tearDown(self):
        shutil.rmtree(self.anonymous_dir)

if __name__ == '__main__':
    unittest.main()


scripts/dianon.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import argparse
import dicom
import sys

if __name__ == "__main__": #pragma nocover

    from dianonymous.dianonymous import anonymize
    parser = argparse.ArgumentParser(description="Anonymize DICOM files")

    parser.add_argument(
        '-r', '--recurse',
        default=False,
        action="store_true",
        help="If input is a directory all DICOM files including"
            "subdirectories will be anonymized"
    )

    parser.add_argument(
        '-o', '--output',
        default="./anonymized/",
        help="Directory to place output files. If it doesn't exist it will be created."
    )

    parser.add_argument(
        '-p', '--patient-id',
        default=None,
        help="Anonymous patient id to use"
    )

    parser.add_argument(
        '-n', '--patient-name',
        default=None,
        help="Anonymous patient name to use"
    )

    parser.add_argument( 'files', nargs="+", help="Input files and directories")

    args = parser.parse_args()

    anonymize(args.files, output=args.output, anon_id=args.patient_id, anon_name=args.patient_name, recurse=args.recurse, log=sys.stdout)


dianonymous/dianonymous.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import argparse
import dicom
import hashlib
import os
import sys

from . import anonymizer

PATIENT_ID = (0x0010, 0x0020)

def anonymize(inputs, output="anonymized", anon_id=None, anon_name=None, recurse=False, private=True, curves=True, log=None):
    """
    Anonymize DICOM files and write new DICOM files to the output directory.
    The original files will be untouched.

    inputs: An iterable of file and directory names.  If one of the inputs
    is a directory, all DICOM files in that directory will be anonymized.

    output: Directory to place output files. If it doesn't exist, it will be
    created.

    recurse: Recurse over all subdirectoies. Default is False.

    private: Anonymize all private tags. Default is True.

    curves: Remove all curves. Default is True.

    log: object with a .write method or None. Default is None

    """

    output_dir = get_output_dir(output)
    completed = []

    for inp in get_input_paths(inputs, recurse):
        try:
            dcm = dicom.read_file(inp, force=True)
            patient_id = dcm[PATIENT_ID].value or "not provided" # also serves as check for valid DICOM file
            patient_name = dcm.PatientsName or "not provided"
            anon_id = anon_id or hashlib.sha512(patient_id).hexdigest()[:10]
            anon_name = anon_name or hashlib.sha512(patient_name).hexdigest()[:10]

            anonymizer.anonymize(dcm, patient_id=anon_id, patient_name=anon_name)

            rel_path = inp.replace(patient_id, anon_id).replace(patient_name, anon_name)
            rel_path = rel_path.replace(".." + os.path.sep, "").replace("." + os.path.sep, "")
            out_path = os.path.join(output_dir, rel_path)
            out_dir = os.path.split(out_path)[0]
            try:
                os.makedirs(out_dir)
            except OSError:
                pass
            dcm.save_as(out_path)
            if log:
                log.write("Anonymized & wrote %s to %s\n" % (inp, out_path))
        except (MemoryError, KeyError):
            pass


#---------------------------------------------------------------
def get_output_dir(output):
    """
    return full path of directory where outputs are to be stored.

    If the directory doesn't exist an attempt will be made to create it.
    """

    path = os.path.abspath(output)

    try:
        os.mkdir(path)
    except OSError:
        #Directory exists already
        pass

    return path


#---------------------------------------------------------------
def get_input_paths(inputs, recurse):
    """
    a generator for all input files requested
    """

    if isinstance(inputs, basestring):
        inputs = [inputs]

    for inp in inputs:
        path = os.path.relpath(inp)

        if os.path.isdir(path) and recurse:
            all_paths = os.walk(path) if recurse else []
        elif os.path.isdir(path):
            all_paths = [(path,"", [f for f in os.listdir(path) if os.path.isfile(os.path.join(path,f))])]

        else:
            root, fname = os.path.split(path)
            all_paths = [(root, "", [fname])]

        for root, _, files in all_paths:
            for f in files:
                yield os.path.join(root,f)


if __name__ == "__main__": #pragma nocover

    parser = argparse.ArgumentParser(description="Anonymize DICOM files")

    parser.add_argument(
        '-r', '--recurse',
        default=False,
        action="store_true",
        help="If input is a directory all DICOM files including"
            "subdirectories will be anonymized"
    )

    parser.add_argument(
        '-o', '--output',
        default="./anonymized/",
        help="Directory to place output files. If it doesn't exist it will be created."
    )

    parser.add_argument(
        '-p', '--patient-id',
        default=None,
        help="Anonymous patient id to use"
    )

    parser.add_argument(
        '-n', '--patient-name',
        default=None,
        help="Anonymous patient name to use"
    )

    parser.add_argument( 'files', nargs="+", help="Input files and directories")

    args = parser.parse_args()

    anonymize(args.files, output=args.output, anon_id=args.patient_id, anon_name=args.patient_name, recurse=args.recurse, log=sys.stdout)


dianonymous/anonymizer.py
import dicom
import uuid
import hashlib


ANONYMOUS_AGE = "200"
AGE_STRING = "AS"
PERSONS_NAME = "PN"

UPPER_BYTE_MASK = 0xFF00
UPPER_BYTE_OF_CURVE_GROUP = 0x5000

def anonymous_id(ds, elem):
    return "1234567890"

def anonymous_study_id(ds, elem):
    return "1"

def anonymous_uid(ds, elem):
    return uuid.UUID(hashlib.sha512(str(elem.tag)).hexdigest()[:32]).get_hex()

def anonymous_string(ds, elem):
    return "anonymous"

def anonymous_date(ds, elem):
    return "19010101"

def anonymous_datetime(ds, elem):
    return "19010101000000.000000"

def anonymous_time(ds, elem):
    return "000000.000"

def anonymous_number(ds, elem):
    return 0

def anonymous_sex(ds, elem):
    return ""

def anonymous_decimal_string(ds, elem):
    return "0.000"

def anonymous_integer_string(ds, elem):
    return "0"

def no_change(ds, elem):
    return elem.value


DEFAULT_METHODS = {
    "AE" : no_change,
    "AS" : anonymous_string,
    "AT" : no_change,
    "CS" : no_change,
    "DA" : anonymous_date,
    "DS" : no_change,
    "DT" : anonymous_datetime,
    "FL" : no_change,
    "FD" : no_change,
    "IS" : no_change,
    "LO" : no_change,
    "LT" : no_change,
    "OB" : no_change,
    "OF" : no_change,
    "OW" : no_change,
    "PN" : anonymous_string,
    "SH" : no_change,
    "SL" : no_change,
    "SQ" : no_change,
    "SS" : no_change,
    "ST" : no_change,
    "TM" : anonymous_string,
    "UI" : no_change,
    "UL" : no_change,
    "UN" : no_change,
    "US" : no_change,
    "UT" : no_change,
}


TAGS_TO_ANONYMIZE = {
    (0x0008, 0x0014): {"name": "InstanceCreatorUID", "method": anonymous_uid},
    (0x0008, 0x0018): {"name": "SOPInstanceUID", "method": anonymous_uid},
    (0x0008, 0x0050): {"name": "AccessionNumber", "method": None},
    (0x0008, 0x0080): {"name": "InstitutionName", "method": anonymous_string},
    (0x0008, 0x0081): {"name": "InstitutionAddress", "method": anonymous_string},
    (0x0008, 0x0090): {"name": "ReferringPhysiciansName", "method": anonymous_string},
    (0x0008, 0x0092): {"name": "ReferringPhysiciansAddress", "method": anonymous_string},
    (0x0008, 0x0094): {"name": "ReferringPhysiciansTelephoneNumber", "method": anonymous_string},
    (0x0008, 0x1010): {"name": "StationName", "method": anonymous_string},
    (0x0008, 0x1030): {"name": "StudyDescription", "method": anonymous_string},
    (0x0008, 0x103E): {"name": "SeriesDescription", "method": anonymous_string},
    (0x0008, 0x1040): {"name": "InstitutionalDepartmentName", "method": anonymous_string},
    (0x0008, 0x1048): {"name": "PhysiciansofRecord", "method": anonymous_string},
    (0x0008, 0x1050): {"name": "PerformingPhysiciansName", "method": anonymous_string},
    (0x0008, 0x1060): {"name": "NameofPhysiciansReadingStudy", "method": anonymous_string},
    (0x0008, 0x1070): {"name": "OperatorsName", "method": anonymous_string},
    (0x0008, 0x1080): {"name": "AdmittingDiagnosesDescription", "method": None},
    (0x0008, 0x1155): {"name": "ReferencedSOPInstanceUID", "method": anonymous_uid},
    (0x0008, 0x2111): {"name": "DerivationDescription", "method": None},
    (0x0010, 0x0010): {"name": "PatientsName", "method": anonymous_string},
    (0x0010, 0x0020): {"name": "PatientID", "method": anonymous_id},
    (0x0010, 0x0030): {"name": "PatientsBirthDate", "method": anonymous_date},
    (0x0010, 0x0032): {"name": "PatientsBirthTime", "method": anonymous_time},
    (0x0010, 0x0040): {"name": "PatientsSex", "method": anonymous_sex},
    (0x0010, 0x1000): {"name": "OtherPatientIds", "method": anonymous_string},
    (0x0010, 0x1001): {"name": "OtherPatientNames", "method": anonymous_string},
    (0x0010, 0x1010): {"name": "PatientsAge", "method": anonymous_integer_string},
    (0x0010, 0x1020): {"name": "PatientsSize", "method": anonymous_number},
    (0x0010, 0x1030): {"name": "PatientsWeight", "method": anonymous_decimal_string},
    (0x0010, 0x1040): {"name": "PatientsAddress", "method": anonymous_string},
    (0x0010, 0x1090): {"name": "MedicalRecordLocator", "method": None},
    (0x0010, 0x2160): {"name": "EthnicGroup", "method": anonymous_string},
    (0x0010, 0x2180): {"name": "Occupation", "method": anonymous_string},
    (0x0010, 0x21B0): {"name": "AdditionalPatientsHistory", "method": anonymous_string},
    (0x0010, 0x4000): {"name": "PatientComments", "method": anonymous_string},
    (0x0018, 0x1000): {"name": "DeviceSerialNumber", "method": anonymous_string},
    (0x0018, 0x1030): {"name": "ProtocolName", "method": anonymous_string},
    (0x0020, 0x000D): {"name": "StudyInstanceUID", "method": anonymous_uid},
    (0x0020, 0x000E): {"name": "SeriesInstanceUID", "method": anonymous_uid},
    (0x0020, 0x0010): {"name": "StudyID", "method": anonymous_study_id},
    (0x0020, 0x0052): {"name": "FrameofReferenceUID", "method": anonymous_uid},
    (0x0020, 0x0200): {"name": "SynchronizationFrameofReferenceUID", "method": anonymous_uid},
    (0x0020, 0x4000): {"name": "ImageComments", "method": anonymous_string},
    (0x0040, 0x0275): {"name": "RequestAttributesSequence", "method": None},
    (0x0040, 0xA124): {"name": "UID", "method": anonymous_uid},
    (0x0040, 0xA730): {"name": "ContentSequence", "method": None},
    (0x0088, 0x0140): {"name": "StorageMediaFile-setUID", "method": anonymous_uid},
    (0x3006, 0x0024): {"name": "ReferencedFrameofReferenceUID ", "method": anonymous_uid},
    (0x3006, 0x00C2): {"name": "RelatedFrameofReferenceUID", "method": anonymous_uid},
}


SET_OF_TAGS_TO_ANONYMIZE = set(TAGS_TO_ANONYMIZE.keys())

def anonymize(dcm, patient_id=None, patient_name=None, private=True ):
    """return an anonymized version of input dcm file object """

    # Define call-back functions for the dcm.walk() function
    def callback(ds, data_element):
        """Called from the dcm "walk" recursive function for all data elements."""

        t = data_element.tag
        hext = (t.group, t.elem)

        if patient_name is not None and hext in ((0x0010, 0x0010), (0x0010, 0x1001),):
            data_element.value = patient_name
        elif patient_id is not None and hext in ((0x0010, 0x1000), (0x0010, 0x0020),):
            data_element.value = patient_id
        elif private and t.is_private:
            del ds[t]
        elif hext in SET_OF_TAGS_TO_ANONYMIZE:
            fun = TAGS_TO_ANONYMIZE[hext]["method"]
            if fun is None:
                fun = DEFAULT_METHODS[data_element.VR]
            try:
                data_element.value = fun(ds, data_element)
            except ValueError:
                raise ValueError("Unable to convert data_element (%s, %s) with function %s" % (data_element.tag, data_element.name, fun))
        elif t.group & UPPER_BYTE_MASK == UPPER_BYTE_OF_CURVE_GROUP:
            del ds[t]


    dcm.walk(callback)
